File: game-engine/src/test/java/org/luxons/sevenwonders/game/cards/CardTest.java
package org.luxons.sevenwonders.game.cards;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.junit.Before;
import org.junit.Test;
import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.effects.Effect;
import org.luxons.sevenwonders.game.effects.ProductionIncrease;
import org.luxons.sevenwonders.game.resources.ResourceTransactions;
import org.luxons.sevenwonders.game.resources.ResourceType;
import org.luxons.sevenwonders.game.test.TestUtilsKt;
import org.luxons.sevenwonders.game.wonders.Wonder;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;

public class CardTest {

    private Table table;

    private Card treeFarmCard;

    @Before
    public void initBoard() {
        Settings settings = new Settings(3);

        List<Board> boards = new ArrayList<>(3);
        boards.add(new Board(new Wonder("TestWonder", ResourceType.WOOD), 0, settings));
        boards.add(new Board(new Wonder("TestWonder", ResourceType.STONE), 1, settings));
        boards.add(new Board(new Wonder("TestWonder", ResourceType.PAPYRUS), 2, settings));
        table = new Table(boards);

        Requirements treeFarmRequirements = new Requirements();
        treeFarmRequirements.setGold(1);

        ProductionIncrease treeFarmEffect = new ProductionIncrease();
        treeFarmEffect.getProduction().addChoice(ResourceType.WOOD, ResourceType.CLAY);

        List<Effect> effects = Collections.singletonList(treeFarmEffect);

        treeFarmCard = new Card("Tree Farm", Color.BROWN, treeFarmRequirements, effects, "", null, null);
    }

    @Test
    public void playCardCostingMoney() {
        table.getBoard(0).setGold(3);
        table.getBoard(1).setGold(3);
        table.getBoard(2).setGold(3);
        treeFarmCard.applyTo(table, 0, new ResourceTransactions());
        assertEquals(2, table.getBoard(0).getGold());
        assertEquals(3, table.getBoard(1).getGold());
        assertEquals(3, table.getBoard(2).getGold());
    }

    @Test
    public void equals_falseWhenNull() {
        Card card = TestUtilsKt.testCard("TestCard");
        //noinspection ObjectEqualsNull
        assertFalse(card.equals(null));
    }

    @Test
    public void equals_falseWhenDifferentClass() {
        Card card = TestUtilsKt.testCard("TestCard");
        Object object = new Object();
        //noinspection EqualsBetweenInconvertibleTypes
        assertFalse(card.equals(object));
    }

    @Test
    public void equals_trueWhenSame() {
        Card card = TestUtilsKt.testCard("TestCard");
        assertEquals(card, card);
    }

    @Test
    public void equals_trueWhenSameContent() {
        Card card1 = TestUtilsKt.testCard("TestCard");
        Card card2 = TestUtilsKt.testCard("TestCard");
        assertTrue(card1.equals(card2));
    }

    @Test
    public void equals_falseWhenDifferentName() {
        Card card1 = TestUtilsKt.testCard("TestCard1");
        Card card2 = TestUtilsKt.testCard("TestCard2");
        assertFalse(card1.equals(card2));
    }

    @Test
    public void hashCode_sameWhenSameContent() {
        Card card1 = TestUtilsKt.testCard("TestCard");
        Card card2 = TestUtilsKt.testCard("TestCard");
        assertEquals(card1.hashCode(), card2.hashCode());
    }

    @Test
    public void hashCode_differentWhenDifferentName() {
        Card card1 = TestUtilsKt.testCard("TestCard1");
        Card card2 = TestUtilsKt.testCard("TestCard2");
        assertNotEquals(card1.hashCode(), card2.hashCode());
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/cards/CardTest.kt
package org.luxons.sevenwonders.game.cards

import org.junit.Test
import org.luxons.sevenwonders.game.SimplePlayer
import org.luxons.sevenwonders.game.api.Table
import org.luxons.sevenwonders.game.boards.Board
import org.luxons.sevenwonders.game.effects.ProductionIncrease
import org.luxons.sevenwonders.game.resources.Production
import org.luxons.sevenwonders.game.resources.ResourceType
import org.luxons.sevenwonders.game.resources.noTransactions
import org.luxons.sevenwonders.game.test.testCard
import org.luxons.sevenwonders.game.test.testSettings
import org.luxons.sevenwonders.game.wonders.Wonder
import kotlin.test.assertEquals

class CardTest {

    @Test
    fun playCardCostingMoney() {
        val initialGold = 3
        val price = 1
        val settings = testSettings(3, initialGold)

        val boards = listOf(
            Board(Wonder("TestWonder", ResourceType.WOOD, emptyList(), ""), 0, settings),
            Board(Wonder("TestWonder", ResourceType.STONE, emptyList(), ""), 1, settings),
            Board(Wonder("TestWonder", ResourceType.PAPYRUS, emptyList(), ""), 2, settings)
        )
        val table = Table(boards)

        val treeFarmRequirements = Requirements(gold = price)
        val treeFarmProduction = Production().apply { addChoice(ResourceType.WOOD, ResourceType.CLAY) }
        val treeFarmEffect = ProductionIncrease(treeFarmProduction, false)
        val treeFarmCard = testCard("Tree Farm", Color.BROWN, treeFarmRequirements, treeFarmEffect)

        treeFarmCard.applyTo(SimplePlayer(0, table), noTransactions())

        assertEquals(initialGold - price, table.getBoard(0).gold)
        assertEquals(initialGold, table.getBoard(1).gold)
        assertEquals(initialGold, table.getBoard(2).gold)
    }
}
